src/mpid/ch4/src/ch4_wait.h

```diff
@@ -8,57 +8,125 @@
 
 #include "ch4_impl.h"
 
+MPL_STATIC_INLINE_PREFIX void MPIDI_set_progress_vci(MPIR_Request * req,
+                                                     MPID_Progress_state * state)
+{
+    state->flag = MPIDI_PROGRESS_ALL;   /* TODO: check request is_local/anysource */
+    state->progress_count = MPL_atomic_load_int(&MPIDI_global.progress_count);
+    state->progress_start = state->progress_count;
+
+    state->vci_count = 1;
+    state->vci[0] = (req->handle & REQUEST_POOL_MASK) >> REQUEST_POOL_SHIFT;
```

Comment:
So pool_idx is equal as vci_idx here, assuming number of pools == number of vcis, right?

Reply:
It is strictly `num_pools >= num_vcis >= num_vnis`. The request is allocated by lower layers, so it will always fit into upper-layer's allocations. Greater is allowed, so for example, we may let `netmod` and `shmmod` occupy two disjoint ch4 vci space.

Reply:
I must miss the code of checking `vci_idx` > `num_vcis` in this case, then how you handle it?

Reply:
> 
> 
> I must miss the code of checking `vci_idx` > `num_vcis` in this case, then how you handle it?

It is in development :). Currently it is strictly equal: https://github.com/pmodels/mpich/blob/55c2deeadfb5aedb958f281729dcf678cb24aff2/src/mpid/ch4/netmod/ofi/ofi_init.c#L572

We'll have to update it as we add vci support to shm and am system.